cmd/controller/app/start.go

```diff
@@ -38,64 +45,169 @@ import (
 	_ "github.com/cert-manager/cert-manager/pkg/issuer/venafi"
 	logf "github.com/cert-manager/cert-manager/pkg/logs"
 	"github.com/cert-manager/cert-manager/pkg/util"
+	"github.com/cert-manager/cert-manager/pkg/util/configfile"
 	utilfeature "github.com/cert-manager/cert-manager/pkg/util/feature"
 )
 
-type CertManagerControllerOptions struct {
-	ControllerOptions *options.ControllerOptions
-}
+const componentController = "controller"
 
-func NewCertManagerControllerOptions() *CertManagerControllerOptions {
-	o := &CertManagerControllerOptions{
-		ControllerOptions: options.NewControllerOptions(),
-	}
+func NewServerCommand(stopCh <-chan struct{}) *cobra.Command {
+	ctx := cmdutil.ContextWithStopCh(context.Background(), stopCh)
+	log := logf.Log
 
-	return o
-}
+	ctx = logf.NewContext(ctx, log, componentController)
 
-// NewCommandStartCertManagerController is a CLI handler for starting cert-manager
-func NewCommandStartCertManagerController(stopCh <-chan struct{}) *cobra.Command {
-	o := NewCertManagerControllerOptions()
+	cleanFlagSet := pflag.NewFlagSet(componentController, pflag.ContinueOnError)
+	// Replaces all instances of `_` in flag names with `-`
+	cleanFlagSet.SetNormalizeFunc(cliflag.WordSepNormalizeFunc)
+	controllerFlags := options.NewControllerFlags()
+	controllerConfig, err := options.NewControllerConfiguration()
+	if err != nil {
+		log.Error(err, "Failed to create new controller configuration")
+		os.Exit(1)
+	}
 
 	cmd := &cobra.Command{
-		Use:   "cert-manager-controller",
+		Use:   componentController,
 		Short: fmt.Sprintf("Automated TLS controller for Kubernetes (%s) (%s)", util.AppVersion, util.AppGitCommit),
 		Long: `
 cert-manager is a Kubernetes addon to automate the management and issuance of
 TLS certificates from various issuing sources.
 
 It will ensure certificates are valid and up to date periodically, and attempt
 to renew certificates at an appropriate time before expiry.`,
+		// The controller has special flag parsing requirements to handle precedence of providing
+		// configuration via versioned configuration files and flag values.
+		// Setting DisableFlagParsing=true prevents Cobra from interfering with flag parsing
+		// at all, and instead we handle it all in the RunE below.
```

Comment:
It seems quite unfortunate that we have to workaround here.
Do we have the same workaround for the webhook?
Could we maybe create some kind of wrapper/ util function that provides this workaround?

Reply:
I pulled this code mostly from the webhook example, so yes we have the same workaround. We _could_ create a wrapper around this. But I'm tempted to file a separate issue to track this. What are your thoughts on that?

Reply:
Also, for reference, this is how it is done in the kubelet code: https://github.com/kubernetes/kubernetes/blob/f61a0e41075c2c4468e6cda84f7a88a960fb2933/cmd/kubelet/app/server.go#L168-L304

Reply:
I think I found an easier way to do this, material for a follow-up PR.